File: backend/internal/domain/content.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// ContentStatus represents the production lifecycle stage of a piece of content.
type ContentStatus string

const (
	ContentStatusIdea      ContentStatus = "idea"
	ContentStatusScripting ContentStatus = "scripting"
	ContentStatusRecording ContentStatus = "recording"
	ContentStatusEditing   ContentStatus = "editing"
	ContentStatusReview    ContentStatus = "review"
	ContentStatusScheduled ContentStatus = "scheduled"
	ContentStatusPublished ContentStatus = "published"
	ContentStatusArchived  ContentStatus = "archived"
)

// validTransitions defines the allowed state machine transitions for content status.
var validTransitions = map[ContentStatus][]ContentStatus{
	ContentStatusIdea:      {ContentStatusScripting, ContentStatusArchived},
	ContentStatusScripting: {ContentStatusRecording, ContentStatusIdea, ContentStatusArchived},
	ContentStatusRecording: {ContentStatusEditing, ContentStatusScripting, ContentStatusArchived},
	ContentStatusEditing:   {ContentStatusReview, ContentStatusRecording, ContentStatusArchived},
	ContentStatusReview:    {ContentStatusScheduled, ContentStatusEditing, ContentStatusArchived},
	ContentStatusScheduled: {ContentStatusPublished, ContentStatusReview, ContentStatusArchived},
	ContentStatusPublished: {ContentStatusArchived},
	ContentStatusArchived:  {},
}

// ValidateStatusTransition checks whether transitioning from `from` to `to` is permitted.
// Returns a domain AppError if the transition is invalid.
func ValidateStatusTransition(from, to ContentStatus) error {
	allowed, ok := validTransitions[from]
	if !ok {
		return NewError("CONTENT_001", "unknown source status", 400)
	}
	for _, s := range allowed {
		if s == to {
			return nil
		}
	}
	return NewError("CONTENT_002", "invalid status transition", 400)
}

// ContentType categorises content by format.
type ContentType string

const (
	ContentTypeVideo  ContentType = "video"
	ContentTypeShort  ContentType = "short"
	ContentTypeReel   ContentType = "reel"
	ContentTypePost   ContentType = "post"
	ContentTypeThread ContentType = "thread"
)

// Content is the core content aggregate.
type Content struct {
	ID             uuid.UUID
	WorkspaceID    uuid.UUID
	CreatedBy      uuid.UUID
	Title          string
	Description    *string
	Status         ContentStatus
	ContentType    ContentType
	PlatformTarget *PlatformType
	ScheduledAt    *time.Time
	PublishedAt    *time.Time
	DueDate        *time.Time
	Metadata       map[string]any
	CreatedAt      time.Time
	UpdatedAt      time.Time
	DeletedAt      *time.Time
}

// ContentAssignment links a user to a content item with a role.
type ContentAssignment struct {
	ID         uuid.UUID
	ContentID  uuid.UUID
	UserID     uuid.UUID
	Role       string
	AssignedAt time.Time
}

// ContentFilter holds optional filtering parameters for listing content.
type ContentFilter struct {
	Statuses   []ContentStatus
	AssignedTo *uuid.UUID
	Platform   *PlatformType
	GroupBy    string // "status" for kanban
	Limit      int
	Offset     int
}

// KanbanBoard groups content by status column.
type KanbanBoard struct {
	Columns map[string][]*Content `json:"columns"`
}

File: frontend/packages/types/src/content.ts
import type { UUID, Timestamp } from './common';

export type ContentStatus =
  | 'idea'
  | 'scripting'
  | 'recording'
  | 'editing'
  | 'review'
  | 'scheduled'
  | 'published'
  | 'archived';

export type ContentType = 'video' | 'short' | 'reel' | 'post' | 'thread';

export type PlatformType = 'youtube' | 'tiktok' | 'instagram' | 'twitter' | 'linkedin';

/** @deprecated Use ContentStatus instead. Kept for backward compatibility. */
export type PipelineStage = 'idea' | 'scripting' | 'recording' | 'editing' | 'review' | 'publishing';

export interface ContentItem {
  id: UUID;
  workspace_id: UUID;
  created_by: UUID;
  title: string;
  description?: string | null;
  status: ContentStatus;
  content_type: ContentType;
  platform_target?: PlatformType | null;
  scheduled_at?: Timestamp | null;
  published_at?: Timestamp | null;
  due_date?: Timestamp | null;
  metadata?: Record<string, unknown> | null;
  created_at: Timestamp;
  updated_at: Timestamp;
  deleted_at?: Timestamp | null;

  // Legacy fields kept for backward compatibility
  /** @deprecated Use description instead */
  body?: string | null;
  /** @deprecated Use content_type and platform_target instead */
  platform?: string;
  /** @deprecated Use ContentStatus instead */
  pipeline_stage?: PipelineStage;
  idea_id?: UUID | null;
  thumbnail_url?: string | null;
  tags?: string[];
}

export interface ContentAssignment {
  id: UUID;
  content_id: UUID;
  user_id: UUID;
  role: string;
  assigned_at: Timestamp;
}

export interface KanbanBoard {
  columns: Record<string, ContentItem[]>;
}

export interface Series {
  id: UUID;
  workspace_id: UUID;
  created_by: UUID;
  title: string;
  description?: string | null;
  platform?: PlatformType | null;
  template?: Record<string, unknown> | null;
  created_at: Timestamp;
  updated_at: Timestamp;
  deleted_at?: Timestamp | null;
}

export interface SeriesEpisode {
  id: UUID;
  series_id: UUID;
  content_id?: UUID | null;
  episode_number: number;
  title: string;
  status: ContentStatus;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface SeriesPublishingSchedule {
  id: UUID;
  series_id: UUID;
  frequency: string; // daily, weekly, biweekly, monthly
  day_of_week?: number | null; // 0=Sunday, 6=Saturday
  time_of_day: string; // HH:MM
  timezone: string;
  is_active: boolean;
  created_at: Timestamp;
  updated_at: Timestamp;
}
